### ISSUE TITLE
Default file locations are not correct for Debian 10

### ISSUE THREAD

**Issue by BrainStone** [arkane-systems/genie#69]
I'm not sure which version of Debian you target with the defaults but on my installation of Debian 10 I had to use the Ubuntu deviations file to get everything working.

**Comment by cerebrate** [comment]
At the moment I'm running bullseye (testing/11), but this installation's been upgraded from stretch (9) through buster (10) to get there without changing genie.

Hm. I'm guessing this could be related to the /usr merge. Before I go futzing around with the defaults, would you mind letting me have an ls -la / of your installation, and confirm the presence or absence of /usr/sbin/daemonize, /usr/bin/systemd, /usr/sbin/runuser, and /usr/bin/mount ?

**Comment by BrainStone** [comment]
```
# ls -la
total 704
drwxr-xr-x  22 root root   4096 Jul 25 20:44 .
drwxr-xr-x  22 root root   4096 Jul 25 20:44 ..
drwxr-xr-x   2 root root   4096 Jun 13 17:40 bin
drwxr-xr-x   2 root root   4096 Jan 22  2019 boot
drwxr-xr-x  11 root root   3220 Jul 25 20:35 dev
drwxr-xr-x  94 root root   4096 Jul 25 20:35 etc
drwxr-xr-x   3 root root   4096 Mar  5  2019 home
-rwxr-xr-x   3 root root 631968 Dec  7  2019 init
drwxr-xr-x  11 root root   4096 Jul 21 21:09 lib
drwxr-xr-x   2 root root   4096 Apr 15 01:52 lib64
drwx------   2 root root  16384 Apr 10  2019 lost+found
drwxr-xr-x   2 root root   4096 Feb 12  2019 media
drwxr-xr-x   8 root root   4096 Apr 15 01:35 mnt
drwxr-xr-x   3 root root   4096 Jun 14 00:27 opt
dr-xr-xr-x 147 root root      0 Jul 25 20:35 proc
drwx------  15 root root   4096 Jul 25 20:45 root
drwxr-xr-x  18 root root    520 Jul 25 20:36 run
drwxr-xr-x   2 root root   4096 Jun 13 17:40 sbin
drwxr-xr-x   2 root root   4096 Feb 12  2019 srv
dr-xr-xr-x  11 root root      0 Jul 25 20:35 sys
drwxrwxrwt  13 root root   4096 Jul 25 22:30 tmp
drwxr-xr-x  10 root root   4096 Feb 12  2019 usr
drwxr-xr-x  12 root root   4096 Apr 13 23:37 var
```

And none of these paths exist.  
One thing that may be of interest is that I believe I upgraded it from Debian 9. Didn't think about that initially.

**Comment by cerebrate** [comment]
Interesting. That looks like you haven't installed/gone ahead with usrmerge/[the /usr merge](https://wiki.debian.org/UsrMerge), as otherwise your _/bin_, _/lib_, _/lib64_ and _/sbin_ would all be symlinks. That it was an upgrade from Debian 9 makes this make sense, as usrmerge-by-default came in with stretch, Debian 10.

Which means, I believe, that the root of the problem is that I've gone ahead and used the post-usrmerge paths for executables that don't go through the symlinks, since those come first in $PATH, without considering the non-usrmerge case. But since the old paths still work - at least for now - I could go ahead and use those.

...yeah, that should simplify things. I'll do it in the next release.

**Comment by BrainStone** [comment]
Is there a reason you can't have it detect the paths automatically?

**Comment by cerebrate** [comment]
A similar reason to why I don't just invoke them by filename alone and let $PATH sort it out for me; in that case, the security risk of someone sneaking an identically-named executable into a directory further up $PATH and in the latter of tricking the autodetection into picking up the wrong one. Since I'm running these as root, I want to be as certain as I can that I'm running the correct executable from the system location for it.

(I mean, in theory this shouldn't be an issue because all of WSL is running inside and constrained by a single Windows user session anyway and so you can't get access to anything by cracking root that you couldn't access anyway by using `wsl -u root`. On the other hand, the moment I assume that is the case is the moment someone decides to store a bunch of healthcare records or credit card numbers in a WSL instance running an Internet-facing service, and I get have my very own security incident.

...and I've already seen people using genie to do the "run an Internet-facing service under WSL" bit.)

I suppose a compromise would be to have an install script in the package which would detect if a deviations file were necessary and, if so, add it, but it's not something I've had time (or really, expect to have time in the near future) to work on.

**Comment by BrainStone** [comment]
> A similar reason to why I don't just invoke them by filename alone and let $PATH sort it out for me; in that case, the security risk of someone sneaking an identically-named executable into a directory further up $PATH, and in this one, of tricking the autodetection into picking up the wrong one. Since I'm running these as root, I want to be as certain as I can that I'm running the correct executable from the system location for it.

Why exactly is that an issue that you need to fix?  
This is a general issue with all linux systems and somehow it's not relevant enough that when the exact file location is unknown that using ``$PATH`` is perfectly fine.  
If someone can add a binary in a place that comes before the actual binary they would need to have root access in the first place. I really don't understand why you think you need to have special handling that's more secure on the surface.

**Comment by cerebrate** [comment]
> If someone can add a binary in a place that comes before the actual binary they would need to have root access in the first place.

Er, no, they don't. `su` and `sudo` may reset $PATH to root's, but an suid executable doesn't change the environment at all. (Try it yourself:

```
❯ sudo -s
athena-wsl# echo $PATH
/snap/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
```

versus:

```
❯ cp /bin/sh .
❯ sudo chown root:root sh
❯ sudo chmod u+s sh
❯ ./sh
$ echo $PATH
/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:/home/avatar/.local/bin:/snap/bin:/opt/bin:/mnt/c/Windows/System32:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games
```

) All you have to do is add your home folder to the start of $PATH, put a properly named executable in there, invoke a hypothetical genie-using-non-absolute-filespecs, and that executable will now be running as root. (Which, for that matter, is _why_ su and sudo reset $PATH to root's and take other measures to avoid being an arbitrary-escalation-of-privilege exploit simply by existing, no skill required.)

> Why exactly is that an issue that you need to fix?

Which is why this an issue I need to be mindful of, defensive-programming-wise. As my aged grandmother used to say, "Handing out arbitrary escalation of privilege exploits is rude, young man. Don't do that."

**Comment by BrainStone** [comment]
> su and sudo may reset $PATH to root's

And why can't you do that?

**Comment by cerebrate** [comment]
Actually, I should correct myself. sudo doesn't get root's path, because it doesn't want to leak root's shell dotfiles to everyone who can run sudo. Rather, it uses the secure_path setting from its own config file.

Hrm. Maybe I could work with that sort of arrangement.

**Comment by BrainStone** [comment]
That sounds like a reasonable decision.

**Comment by cerebrate** [comment]
Implemented in 1.25.
